[DOC-2703] feat(snowflake): create snowflake step by step instruction; (#103)
tigergraph/cloud-docs: modules/savanna/modules/graph-development/pages/load-data/load-from-snowflake.adoc

= Load from Snowflake
:experimental:

TigerGraph Savanna allows you to load data directly from Snowflake. This guide will walk you through the steps to configure and load data from Snowflake into your TigerGraph workspace.

== 1) Select Source

> .Once you’ve selected btn:[Snowflake], you will be asked to configure the Snowflake data source.
> . Click on image:Screenshot 2024-04-17 at 9.36.58 PM.png[width=50] to add a new Snowflake data source.

Reviewer: ```suggestion
Once you’ve selected btn:[Snowflake], you will be asked to configure the Snowflake data source.

. Click on image:Screenshot 2024-04-17 at 9.36.58 PM.png[width=50] to add a new Snowflake data source.
```

+
image:gcs-empty-data-source.png[]

. You will need to provide your `Snowflake username`, `Snowflake password`, and `Snowflake URL`.
+
image:snowflake-connection-details.png[]

> . Once you have those configured, you can select one or multiple `Table` and `View` within the Snowflake account.

Reviewer: ```suggestion
. Once you have those configured, you can select one or more `Table` and `View` objects within the Snowflake account.
```

. Click btn:[ Connect ] to fetch the Table(s) and View(s).
+
image:snowflake-tables.png[]

. Click btn:[ Next ] to process the Table(s).

== 2) Configure File

> .This step lets you view the source details.
> . The data loading tool will automatically fetch the sample data.

Reviewer: ```suggestion
This step lets you view the source details.

. The data loading tool will automatically fetch the sample data.
```

+
image::snowflake-sample-data.png[]

> . Once you are satisfied with the samle data, click btn:[ Next ] to proceed.

Reviewer: ```suggestion
. Once you are satisfied with the sample data, click btn:[ Next ] to proceed.
```

== 3) Configure Map

> .If you are loading data into a brand new graph, you will be prompted to let our engine generate a schema and mapping for you. Or you can start from scratch. For more details of schema design please refer to xref:savanna:graph-development:design-schema/index.adoc[Design Schema].
> . Select `Generate the schema only` or `Generate the schema and data mapping`.

Reviewer: ```suggestion
If you are loading data into a brand new graph, you will be prompted to let our engine generate a schema and mapping for you. Or you can start from scratch. For more details of schema design please refer to xref:savanna:graph-development:design-schema/index.adoc[Design Schema].

. Select `Generate the schema only` or `Generate the schema and data mapping`.
```

+
image::Screenshot 2024-04-17 at 5.55.35 PM.png[]
+
[CAUTION]
====
The schema generation feature is still a preview feature. The correctness and efficiency of the resulting graph schema and mapping could vary.
====

. In the `Source` column, you can choose the specific column from the data source that you want to map with the attribute.
+
image::config-mapping-snowflake-1.png[]
+
. Use the `+` button to create a new attribute of the target vertex or edge.
+
image::config-mapping-2.png[width=400]

. Click the btn:[Token Function] button to configure token functions for the selected source. For more details of configuring token functions, please refer to xref:savanna:graph-development:load-data/token-function.adoc[Token Function].

. Click the btn:[Quick Map] button to quickly map the data source headers to the existing schema attributes.
+
image::quick-map-snowflake.png[]
+
    .. The btn:[Map all to target] button aligns existing attribute names with the corresponding data source headers, it won't introduce new attributes.

>     .. The btn:[Map all from source] button not only aligns existing attribute names with the corresponding dat asource headers, but also introduces new attributes based on unmatched data source headers.

Reviewer: ```suggestion
    .. The btn:[Map all from source] button not only aligns existing attribute names with the corresponding data source headers, but also introduces new attributes based on unmatched data source headers.
```

>     .. The following list shows the mapping status of each attribute, you can manually adjust the mapping by checking the box next to the attribute name.

Reviewer: ```suggestion
    .. The following list shows the mapping status of each attribute. You can manually adjust the mapping by checking the box next to the attribute name.
```

. Click btn:[Next] to proceed.

== 4) Confirm

> .This step will let you confirm the changes made to the schema and the data mapping you created to load the data.
> . Simply review the `Schema to be changed` and `Data to be loaded` lists.

Reviewer: ```suggestion
This step will let you confirm the changes made to the schema and the data mapping you created to load the data.

. Simply review the `Schema to be changed` and `Data to be loaded` lists.
```

+
image::confirm-snowflake.png[]
+
[CAUTION]
====

> Please be aware that some schema changes will result in unintentional deletion of the data. Please carefully review the warning message before confirming the loading.

Reviewer: @yulitg2110 I don't like the wording of this warning. 
1. What data are you talking about?  We are configuring loading. We haven't loaded any data yet. How can loading delete data?  The load might not work the way that you want (worse case: you get garbage), but the source data should be unaffected, right?
2. What do you mean by "unintentional"?  Do you mean
 a. The user might have make a mistake in their mapping, so that is why it is unintentional.
 b. If you make a *change* to a mapping, it can have complex consequences. The consequences are predictable, but the user might not understand them, so that is why it is unintentional.
  c. If you make a *change* to a mapping, the consequences are unpredictable. Even an expert cannot know what will happen.
  
I strongly recommend that this warning be rewritten to explain more clearly what you mean.  I realize you are just copying the wording that is already used on the other loading pages. They need to be changed too.

Author: Thank you, @victorleeTG .
1. In savanna, users can simultaneously modify the schema and mapping data on the same page. They don't have to strictly follow the steps from the schema changes to the loading jobs like GST does. Thus, if some data were already loaded into the graph, but the schema changes, the side effect of the schema change jobs may remove the existing data;
2. If the user modifies the PRIMARY ID or PRIMARY KEY of a vertex during the data mapping, the vertex data and the related edge data will be affected. I went through the documentation of GSQL, I found that we mentioned it but unclearly: "Data already stored in the graph and which is not logically part of the change is unaffected." https://docs.tigergraph.com/gsql-ref/4.2/ddl-and-loading/modifying-a-graph-schema#_dos_and_donts_for_schema_change
3. Yes, we will make the warning message more clearly, like if a vertex or edge is removed, the corresponding data and loading jobs will be removed as well. @samzhou2
